### PR TITLE
NP-1442/after approving the approval request, it redirects to the earlier screen

### PR DESCRIPTION
**Pull request by olhabn**
draft pr

I decided to add contextActions to use specific methods for contextMenu if required as in case of https://nanopay.atlassian.net/browse/NP-1442, where as Anna suggests in the comments we need to 'go back' to approval request if we use actions in DetailView but we don't need to 'go back' to approval request if we use context menu actions.

### CODE REVIEW

File: src/foam/nanos/approval/ApprovalRequest.js [hunk 1]
```diff
@@ -570,13 +573,28 @@
           this.notify(this.SUCCESS_REJECTED, '', this.LogLevel.INFO, true);
 
           if ( this.stack.top.length > 0 ) {
-            this.stack.back();
+            if ( ! stackBackIgnore )
+              this.stack.back();
+            // else
```
Comment (olhabn) [inline src/foam/nanos/approval/ApprovalRequest.js]: we need to update table as otherwise we will not be able to observe changes immediately. so I thought maybe we can do that with stack somehow, but it doesn't look good.

we have `this.data.sub('on', function()` in UnstyledTableView line 410 but that's not triggered (not sure if it works as expected)

**Review by kgrgreer — CHANGES_REQUESTED** [review]
I would like to investigate alternatives when I return on Wednesday. I'm concerned about having actions vs contextActions, and would like to look into options of unifying the two, possibly with extra meta-information.

File: src/foam/core/Action.js [hunk 2]
```diff
@@ -294,3 +294,23 @@ foam.CLASS({
     }
   ]
 });
+
+
+/** Add Context Action support to Model. */
+foam.CLASS({
```
Comment (Nauna) [inline src/foam/core/Action.js]: don't think you need this

**Comment by olhabn** [comment]
used isContextMenuEnabled and isDetailViewEnabled instead of ContextMenuActions refinement just as an option for tomorrows discussion

File: src/foam/nanos/approval/ApprovalRequest.js [hunk 3]
```diff
@@ -596,10 +610,16 @@
       },
       code: function(X) {
         var objToAdd = X.objectSummaryView ? X.objectSummaryView : X.summaryView;
+        var onExecuteFunc;
+
+        if ( X.controllerMode && X.controllerMode.equals(foam.u2.ControllerMode.SUMMARY) )
```
Comment (kgrgreer) [inline src/foam/nanos/approval/ApprovalRequest.js]: For Enums, you can use ==, instead of equals(). It's faster and shorter. This works because all enum values are singletons.

File: src/foam/nanos/approval/ApprovalRequest.js [hunk 4]
```diff
@@ -596,10 +610,16 @@
...
+          onExecuteFunc =  this.approveWithMemo.bind(this);
```
Comment (kgrgreer) [inline src/foam/nanos/approval/ApprovalRequest.js]: You don't have to make this change, but since these are used as callbacks/listeners, if you moved them from methods: to listeners: then you wouldn't need to .bind(this) them, because listeners are pre-bound.

File: src/foam/nanos/approval/ApprovalRequest.js (hunk 3 above)
Comment (olhabn) [reply src/foam/nanos/approval/ApprovalRequest.js]: fixed

File: src/foam/nanos/approval/ApprovalRequest.js (hunk 4 above)
Comment (olhabn) [reply src/foam/nanos/approval/ApprovalRequest.js]: fixed

File: src/foam/comics/v2/DAOSummaryView.js [hunk 5]
```diff
@@ -109,6 +109,12 @@ foam.CLASS({
       factory: function() { 
         return () => this.stack.back();
       }
+    },
+    {
+      class: 'Enum',
+      of: 'foam.u2.ControllerMode',
+      name: 'controllerMode',
+      value: foam.u2.ControllerMode.SUMMARY
```
Comment (kgrgreer) [inline src/foam/comics/v2/DAOSummaryView.js]: I think that if we decorate the stack, we no longer need this ControllerMode.

File: src/foam/nanos/approval/ApprovalRequest.js [hunk 6]
```diff
@@ -474,6 +478,15 @@
       class: 'String',
       name: 'approvableHashKey',
       hidden: true
+    },
```
Comment (kgrgreer) [inline src/foam/nanos/approval/ApprovalRequest.js]: I don't think this should be specific to the ApprovalRequest, but the subStack should be installed in the TableView so that actions called from there, either from the context menu or directly, don't cause a stack.back().

File: src/foam/nanos/approval/ApprovalRequest.js [hunk 7]
```diff
@@ -795,5 +775,67 @@
...
+        var approvedApprovalRequest = this.clone();
+        approvedApprovalRequest.status = this.ApprovalStatus.APPROVED;
+        approvedApprovalRequest.memo = memo;
+
+        this.approvalRequestDAO.put(approvedApprovalRequest).then((req) => {
+          this.approvalRequestDAO.cmd(this.AbstractDAO.RESET_CMD);
+          this.finished.pub();
+          this.notify(this.SUCCESS_APPROVED, '', this.LogLevel.INFO, true);
+
+          if ( this.subStack.top.length > 0 ) {
```
Comment (kgrgreer) [inline src/foam/nanos/approval/ApprovalRequest.js]: Do we need this 'if'? The back() method in stack already has:
      isEnabled: function(pos) { return pos > 0; },

File: src/foam/nanos/approval/ApprovalRequest.js [hunk 8]
```diff
@@ -795,5 +775,67 @@
...
+        }, (e) => {
+          this.throwError.pub(e);
+          this.notify(e.message, '', this.LogLevel.ERROR, true);
+        });
+      }
+    }
   ]
 });
+
+foam.CLASS({
```
Comment (kgrgreer) [inline src/foam/nanos/approval/ApprovalRequest.js]: Should be in it's own file, or in an inner-class (classes:) in the class/model that instantiates it, which shouldn't be this one.

File: src/foam/u2/Element.js [hunk 9]
```diff
@@ -38,9 +38,10 @@ foam.ENUM({
   ],
 
   values: [
-    { name: 'CREATE', label: 'Create', modePropertyName: 'createVisibility' },
-    { name: 'VIEW',   label: 'View',   modePropertyName: 'readVisibility'   },
-    { name: 'EDIT',   label: 'Edit',   modePropertyName: 'updateVisibility' }
+    { name: 'CREATE',  label: 'Create',  modePropertyName: 'createVisibility' },
+    { name: 'VIEW',    label: 'View',    modePropertyName: 'readVisibility'   },
+    { name: 'EDIT',    label: 'Edit',    modePropertyName: 'updateVisibility' },
+    { name: 'SUMMARY', label: 'Summary', modePropertyName: 'readVisibility'   }
```
Comment (kgrgreer) [inline src/foam/u2/Element.js]: Not needed.

File: src/foam/nanos/approval/ApprovalRequest.js [hunk 10]
```diff
@@ -795,5 +775,67 @@
...
+  package: 'foam.nanos.approval',
+  name: 'NoBackStackProxy',
+  properties: [
+    'delegate',
+    {
+      name: 'top',
```
Comment (kgrgreer) [inline src/foam/nanos/approval/ApprovalRequest.js]: Does this work? When you call nostackbackproxy.top() it will get the right method, but then when it is called it will be bound to the wrong 'this'.
Make it a method and have it just call delegate.push(args) or delgate.apply(this, arguments).

File: src/foam/nanos/approval/ApprovalRequest.js [hunk 11]
```diff
@@ -795,5 +775,67 @@
...
+          this.notify(e.message, '', this.LogLevel.ERROR, true);
+        });
+      }
+    }
   ]
 });
+
+foam.CLASS({
+  package: 'foam.nanos.approval',
+  name: 'NoBackStackProxy',
```
Comment (kgrgreer) [inline src/foam/nanos/approval/ApprovalRequest.js]: Just name NoBackStack. It isn't common to name decorators Proxy, even though decorators are a form of Proxy.

File: src/foam/u2/Element.js (hunk 9 above)
Comment (olhabn) [reply src/foam/u2/Element.js:44]: how we find out where from the approve method is called from without mode?

File: src/foam/nanos/approval/ApprovalRequest.js (hunk 10 above)
Comment (kgrgreer) [reply src/foam/nanos/approval/ApprovalRequest.js]: My mistake, I thought this was push(), which is what you should implement. Why do you need top?

File: src/foam/nanos/approval/ApprovalRequest.js (hunk 11 above)
Comment (olhabn) [reply src/foam/nanos/approval/ApprovalRequest.js]: I don't need it anymore

File: src/foam/nanos/approval/ApprovalRequest.js [hunk 12]
```diff
@@ -39,7 +39,11 @@
     'foam.dao.AbstractDAO',
     'foam.u2.dialog.Popup',
     'foam.log.LogLevel',
-    'foam.nanos.approval.ApprovalStatus'
+    'foam.nanos.approval.ApprovalStatus',
+    {
```
Comment (kgrgreer) [inline src/foam/nanos/approval/ApprovalRequest.js:43]: Is this still needed?

File: src/foam/nanos/approval/ApprovalRequest.js [hunk 13]
```diff
@@ -39,7 +39,11 @@
...
+      path: 'foam.u2.stack.Stack',
```
Comment (kgrgreer) [inline src/foam/nanos/approval/ApprovalRequest.js]: Why is the long-form needed?

File: src/foam/u2/view/NoBackStack.js [hunk 14]
```diff
@@ -0,0 +1,13 @@
+foam.CLASS({
```
Comment (kgrgreer) [inline src/foam/u2/view/NoBackStack.js:7]: Add copyright header.

File: src/foam/u2/view/UnstyledTableView.js [hunk 15]
```diff
@@ -251,6 +252,15 @@ foam.CLASS({
     {
       name: 'dataChangedFlag',
       class: 'Boolean'
+    },
+    {
+      name: 'subStack',
+      transient: true,
+      flags: ['js'],
```
Comment (kgrgreer) [inline src/foam/u2/view/UnstyledTableView.js]: Why is this needed?

Comment (kgrgreer) [reply src/foam/u2/view/UnstyledTableView.js]: I mean flags:

File: src/foam/u2/view/NoBackStack.js (hunk 14 above)
Comment (olhabn) [reply src/foam/u2/view/NoBackStack.js:7]: fixed

File: src/foam/u2/view/UnstyledTableView.js (hunk 15 above)
Comment (olhabn) [reply src/foam/u2/view/UnstyledTableView.js]: fixed, all of those changes were needed for ApprovalRequest java class

File: src/foam/nanos/approval/ApprovalRequest.js (hunk 13 above)
Comment (olhabn) [reply src/foam/nanos/approval/ApprovalRequest.js]: fixed

File: src/foam/nanos/approval/ApprovalRequest.js (hunk 10 above)
Comment (olhabn) [reply src/foam/nanos/approval/ApprovalRequest.js]: I don't need it anymore, my mistake
